### PR TITLE
Support $PROJECT in smoke test gen.

### PR DESCRIPTION
**Pull request by shinfan**
- Support `$PROJECT_ID` token in the smoke test config
- Add `InitValue` class which adds support of variable names in the init fields

### CODE REVIEW

**Comment by garrettjonesgoogle** [comment]
@michaelbausor could you do the first pass?

File: src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline [hunk 1]
```diff
@@ -31,18 +36,34 @@ public class LibraryServiceSmokeTest {
...
-      System.out.println("ok");
+      Options options = new Options();
+      options.addOption("h", "help", false, "show usage");
+      options.addOption(
+          Option.builder()
+              .longOpt("project_id")
+              .desc("Project id")
+              .hasArg()
+              .argName("PROJECT-ID")
+              .required(true)
```
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline]: Do we want this to be required? Or can it be optional, with a default? Especially in the case that nothing in the smoke test config contains $PROJECT_ID, it is odd for it to be required.

File: src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java [hunk 2]
```diff
@@ -69,11 +72,16 @@ public boolean hasFormattedInitValue() {
...
 
-  private static String parseValueString(String valueString, String stringToHash) {
+  private static InitValue parseValueString(String valueString, String stringToHash) {
+    InitValue initValue = new InitValue(valueString, InitValueType.Literal);
     if (valueString.contains(randomValueToken)) {
       String randomValue = Integer.toString(Math.abs(stringToHash.hashCode()));
       valueString = valueString.replace(randomValueToken, randomValue);
+      initValue = new InitValue(valueString, InitValueType.Literal);
+    } else if (valueString.contains(projectIdToken)) {
+      valueString = projectIdVariableName;
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java]: This replaces the whole string, which is different from the randomValue case. Is it possible to allow the project id to be specified alongside other data? If not, it should be an error to specify a value which contains projectIdToken, but is not equal to projectIdToken, because the extra is lost.

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java [hunk 3]
```diff
@@ -342,13 +345,18 @@ private InitValueView getInitValue(MethodTransformerContext context, InitCodeNod
...
-      List<String> varList, InitValueConfig initValueConfig) {
+      MethodTransformerContext context, List<String> varList, InitValueConfig initValueConfig) {
     List<String> formatFunctionArgs = new ArrayList<>();
     for (String entityName : varList) {
       String entityValue = "\"[" + Name.from(entityName).toUpperUnderscore() + "]\"";
       if (initValueConfig.hasFormattingConfigInitialValues()
           && initValueConfig.getCollectionValues().containsKey(entityName)) {
-        entityValue = initValueConfig.getCollectionValues().get(entityName);
+        InitValue initValue = initValueConfig.getCollectionValues().get(entityName);
+        if (initValue.getType() == InitValueType.Variable) {
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: Use a switch statement

File: src/main/java/com/google/api/codegen/metacode/InitValue.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.metacode;
+
+/** A class which represents the initialized value of a field. */
+public class InitValue {
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: I haven't looked through whether this works, but is it possible for this class to be rolled into InitFieldConfig?

File: src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java (hunk 2 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java]: Done. We should not allow users to specify project value as the random value case.

File: src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline (hunk 1 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline]: Done. Changed to optional.

Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline]: What happens when it isn't set? Does it take a default value, or null? Having a default value seems the most sensible. Or perhaps if projectId is used but is null, there is an error.

**Comment by shinfan** [comment]
@michaelbausor PTAL

File: src/main/java/com/google/api/codegen/metacode/InitValue.java (hunk 4 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: Fair enough.

Why not use AutoValue for this class?

File: src/main/java/com/google/api/codegen/metacode/InitCodeNode.java [hunk 5]
```diff
@@ -271,13 +272,14 @@ private void resolveNamesAndTypes(
...
-        validateValue(type, initValueConfig.getInitialValue());
+        validateValue(type, initValueConfig.getInitialValue().getValue());
       } else if (initValueConfig.isEmpty()
           && type.isPrimitive()
           && !type.isRepeated()
           && valueGenerator != null) {
         String newValue = valueGenerator.getAndStoreValue(type, identifier);
-        initValueConfig = InitValueConfig.createWithValue(newValue);
+        initValueConfig =
+            InitValueConfig.createWithValue(new InitValue(newValue, InitValueType.Literal));
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/metacode/InitCodeNode.java]: There are lots of uses of "new InitValue(...)". What do you think about static factory methods, such as InitValue.createLiteral(String value) and InitValue.createVariable(String variableName), or something similar?

File: src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline (hunk 1 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline]: Can we test this logic in the baseline?

File: src/main/java/com/google/api/codegen/metacode/InitValue.java (hunk 4 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: Done

File: src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline (hunk 1 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline]: We only support one smoke test per interface so we can only test one of two logic in the baseline. I tested the first logic since output of the second logic is mostly a subset of the first one.  The alternative is to add one more interface to the test library but that would double the number of the baseline files so it is probably not worth doing.

File: src/main/java/com/google/api/codegen/metacode/InitValue.java [hunk 6]
```diff
@@ -14,29 +14,32 @@
...
+  public static InitValue createVariable(String variableName) {
+    return new AutoValue_InitValue(variableName, InitValueType.Variable);
   }
 
+  public abstract String getValue();
+
+  public abstract InitValueType getType();
+
   @Override
   public boolean equals(Object o) {
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: Is it necessary to override equals and hashcode?

Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: Yes it is used in the test code:
https://github.com/shinfan/toolkit/blob/8a04b2e9dbaba5e81c3ff23ba516d45a52f9bb23/src/test/java/com/google/api/codegen/metacode/SampleInitCodeTest.java#L171

Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: Yes, but AutoValue provides an implementation. Do we need to override that?

File: src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline (hunk 1 above)
Comment (michaelbausor) [inline src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline]: OK yes, I agree lets not add another interface.

Not suggesting in this PR, but is there a reason we only support one smoke test? One seems enough for most cases, but maybe there will be some edge case where we want to smoke test multiple methods? Or maybe not...

File: src/main/java/com/google/api/codegen/metacode/InitValue.java (hunk 6 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitValue.java]: Ah you are right. I forgot that the class is switched to auto value. Removed.

**Comment by shinfan** [comment]
Rebased and addressed feedback. PTAL.

**Comment by michaelbausor** [comment]
LGTM - let @garrettjonesgoogle take a look

File: src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java [hunk 7]
```diff
@@ -76,7 +76,10 @@ private static InitValueConfig createInitValueConfig(
               .get(fieldConfig.fieldPath())
               .withInitialCollectionValue(fieldConfig.entityName(), fieldConfig.value());
     } else if (fieldConfig.hasSimpleInitValue()) {
-      valueConfig = InitValueConfig.createWithValue(stripQuotes(fieldConfig.value()));
+      InitValue initValue = fieldConfig.value();
+      valueConfig =
+          InitValueConfig.createWithValue(
+              InitValue.create(stripQuotes(initValue.getValue()), initValue.getType()));
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java]: would it make sense for InitValue to do the quote stripping?

File: src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java [hunk 8]
```diff
@@ -22,23 +22,26 @@
  */
 @AutoValue
 public abstract class InitFieldConfig {
+  public static final String projectIdVariableName = "project_id";
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java]: make all of these `static final` fields UPPER_SNAKE - they are constants.

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java [hunk 9]
```diff
@@ -139,12 +146,15 @@ private SmokeTestClassView createSmokeTestClassView(SurfaceTransformerContext co
...
+    TestMethodView testMethodView = createSmokeTestMethodView(methodContext);
+
     testClass.apiSettingsClassName(namer.getApiSettingsClassName(service));
     testClass.apiClassName(namer.getApiWrapperClassName(service));
     testClass.name(name);
     testClass.outputPath(namer.getSourceFilePath(outputPath, name));
     testClass.templateFileName(SMOKE_TEST_TEMPLATE_FILE);
-    testClass.method(createSmokeTestMethodView(methodContext));
+    testClass.method(testMethodView);
+    testClass.requireProjectID(requireProjectId(testMethodView.initCode(), context.getNamer()));
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java]: We should be consistent with the casing - I think per Java convention, we should have "Id" (not "ID")

File: src/main/resources/com/google/api/codegen/java/smoke_test.snip [hunk 10]
```diff
@@ -10,21 +10,55 @@
     public static void main(String args[]) {
       Logger.getLogger("").setLevel(Level.WARNING);
       try {
-        executeNoCatch(args);
-        System.out.println("ok");
+        {@testBody(smokeTest)}
+        System.out.println("OK");
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/smoke_test.snip]: You didn't like my lower case "ok"? :-(

File: src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java (hunk 7 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/FieldStructureParser.java]: The quote stripping is used at multiple places in FieldStructureParser so I would prefer to keep it to this class.

File: src/main/resources/com/google/api/codegen/java/smoke_test.snip (hunk 10 above)
Comment (shinfan) [inline src/main/resources/com/google/api/codegen/java/smoke_test.snip]: Ummm in my opinion upper case "OK" is more eye-catching in the output so it would be more readable. But I am OK with both :8ball:

File: src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java (hunk 8 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitFieldConfig.java]: Done

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java (hunk 9 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java]: Done

**Comment by garrettjonesgoogle** [comment]
LGTM
